contrib/http_util/README.md

```diff
@@ -32,6 +32,180 @@ compile 'io.opencensus:opencensus-api:0.12.2'
+      }
+    };
+TextFormat.Getter<HttpRequest> myTextFormatGetter =
+    new TextFormat.Getter<HttpRequest>() {
+      @Override
+      public String get(HttpRequest carrier, String key) {
+        return carrier.getHeader(key);
+      }
+    };
+HttpExtractor<HttpRequest, HttpResponse> extractor =
+    new HttpExtractor<HttpRequest, HttpResponse>() {
+      @Override
+      public Integer getStatusCode(HttpResponse response) {
+        return response.getStatusCode();
+      }
+
+      // other methods that need to be overridden
+      // ...
+    };
+```
+
+### customization for span behaviors
+
+Users can implement `HttpSpanCustomizer` to customize how spans are created and what operations
+need to be done upon starting and ending of the span. This class also provides a default mapping
+from HTTP status code to OpenCensus span Status, and users can override to use their own parsing logic.
+
+Users can enable context propagation by specifying a valid `TextFormat`, which is used for
+context serialization/deserialization.
+
+Below is an example of how the customization for span behaviors should be done:
+
+```java
+// Use B3Format for propagation. You can also use other formats provided in
+// HttpPropagationUtil.
+TextFormat myTextFormat = Tracing.getPropagationComponent().getB3Format();
+HttpSpanCustomizer<HttpRequest, HttpResponse> customizer =
+    new HttpSpanCustomizer<HttpRequest, HttpResponse> {
+      @Override
+      public String getSpanName(
+          HttpRequest request,
+          HttpExtractor<HttpRequest, HttpResponse> extractor) {
+        // user-defined span name
+        return extractor.getPath() + " " + extractor.getMethod(request);
+      }
+
+      @Override
+      public SpanBuilder customizeSpanBuilder(
+          HttpRequest request,
+          SpanBuilder spanBuilder,
+          HttpExtractor<HttpRequest, HttpResponse> extractor) {
+        // do not sample API for heartbeat.
+        if ("/heartbeat".equals(extractor.getPath())) {
+          spanBuilder.setSampler(Samplers.neverSample());
+        }
+        // always record spans locally.
+        return spanBuilder.setRecordEvents(true);
+      }
+
+      // other methods that need to be overridden.
+      // ...
+    };
+// The OpenCensus tracer. You can mock it for testing.
+Tracer tracer = Tracing.getTracer();
+```
+
+### client
+
+Users can create a `HttpClientHandler` to help instrument client-side HTTP request/response.
+
+An example usage of the handler would be:
+
+```java
+HttpClientHandler<HttpRequest, HttpResponse> handler =
+    new HttpClientHandler<HttpRequest, HttpResponse>(tracer, myTextFormat, extractor, customizer);
+
+// Use #handleStart in client to start a new span.
+Span span = handler.handleStart(myTextFormatSetter, request, request);
+HttpResponse response = null;
+Throwable error = null;
+try {
+  // Do something to send the request, and get response code from the server
+  int responseCode = request.getResponseCode();
```

Comment:
remove this as it is confusing :) request usually doesn't have response code.

maybe use real servlet code in this example instead? Doesn't have to be async.. rather it helps it be more concrete.

Here's a chopped example of brave servlet filter that only works for async, but yeah can be used.
```java
    Span span = handler.handleReceive(extractor, httpRequest);
    Throwable error = null;
    try (Tracer.SpanInScope ws = tracer.withSpanInScope(span)) {
      chain.doFilter(httpRequest, httpResponse);
    } catch (IOException | ServletException | RuntimeException | Error e) {
      error = e;
      throw e;
    } finally {
        handler.handleSend(ADAPTER.adaptResponse(httpRequest, httpResponse), error, span);
   }
```

Reply:
Thanks! The `getResponseCode` comes from `HttpURLConnection` and I think I kind of mixed them up.

I've written a small netty example here: https://github.com/census-instrumentation/opencensus-experiments/blob/master/integration/src/main/java/io/opencensus/interop/http/netty/OpenCensusClientInboundOutboundHandler.java